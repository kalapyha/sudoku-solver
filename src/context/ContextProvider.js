import React, { useState, useContext } from 'react';
import {
	defaultBaord_9by9,
	emptyBaord_9by9,
	solveSudoku,
} from '../utils/sudoku_solver';

const LoadingContext = React.createContext();
const BoardToSolveContext = React.createContext();
const ResultBoardContext = React.createContext();
const SudokuSolveContext = React.createContext();

export function useLoading() {
	return useContext(LoadingContext);
}

export function useBoardToSolve() {
	return useContext(BoardToSolveContext);
}

export function useResultBoard() {
	return useContext(ResultBoardContext);
}

export function useSolveBoard() {
	return useContext(SudokuSolveContext);
}

export function ContextProvider({ children }) {
	const [isLoading, setisLoading] = useState(false);
	const [boardToSolve, setboardToSolve] = useState(defaultBaord_9by9);
	const [resultBoard, setResultBoard] = useState(emptyBaord_9by9);

	function handleSolve() {
		setisLoading(true);
		const { board, time } = solveSudoku(boardToSolve);
		setResultBoard(board);

		setisLoading(false);
		console.log(`Time spent to solve Sudoku: ${time}ms`);
	}
	return (
		<LoadingContext.Provider value={isLoading}>
			<BoardToSolveContext.Provider value={boardToSolve}>
				<ResultBoardContext.Provider value={resultBoard}>
					<SudokuSolveContext.Provider value={handleSolve}>
						{children}
					</SudokuSolveContext.Provider>
				</ResultBoardContext.Provider>
			</BoardToSolveContext.Provider>
		</LoadingContext.Provider>
	);
}
