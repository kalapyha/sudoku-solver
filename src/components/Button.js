import React from 'react';
import { useGlobalContext } from '../context/ContextProvider';

function Button() {
	const { solveSudokuBoard, isBtnDisabled } = useGlobalContext();

	return (
		<button
			className="btn btn-success"
			onClick={solveSudokuBoard}
			disabled={isBtnDisabled && true}
		>
			Solve!
		</button>
	);
}

export default Button;
