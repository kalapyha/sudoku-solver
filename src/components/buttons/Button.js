import React from 'react';
import { useGlobalContext } from '../../context/ContextProvider';

function Button() {
	const { isBtnDisabled, startLoading, startSolving } = useGlobalContext();

	const handleClick = () => {
		startLoading();
		startSolving();
	};

	return (
		<button
			data-testid="button"
			className="btn btn-primary"
			onClick={handleClick}
			disabled={isBtnDisabled}
		>
			Solve!
		</button>
	);
}

export default Button;
