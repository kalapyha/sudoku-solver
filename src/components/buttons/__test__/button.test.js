import React from 'react';
import { AppProvider } from '../../../context/ContextProvider';
import { render, screen, fireEvent } from '@testing-library/react';
import TestRenderer from 'react-test-renderer';
import Button from '../Button';

beforeEach(() => {
	const div = document.createElement('div');
	render(
		<AppProvider>
			<Button />
		</AppProvider>,
		div
	);
});

describe('<Button />', () => {
	it('renders without crushing', () => {
		const button = screen.getByTestId('button');
		expect(button).toBeInTheDocument();
	});

	it('renders correctly', () => {
		const button = screen.getByTestId('button');
		expect(button).toHaveTextContent('Solve!');
	});

	it('clicks correctly', () => {
		global.console.log = jest.fn();
		const button = screen.getByTestId('button');
		fireEvent.click(button);
		expect(console.log).toBeCalledTimes(1);
	});

	it('matches snapshot', () => {
		const tree = TestRenderer.create(
			<AppProvider>
				<Button />
			</AppProvider>
		).toJSON();
		expect(tree).toMatchSnapshot();
	});
});
