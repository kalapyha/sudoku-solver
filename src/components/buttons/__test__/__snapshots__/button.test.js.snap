// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> matches snapshot 1`] = `
<button
  className="btn btn-primary"
  data-testid="button"
  disabled={false}
  onClick={[Function]}
>
  Solve!
</button>
`;
